tests/run-make/rustdoc-io-error/rmake.rs

```diff
@@ -0,0 +1,32 @@
+// This test verifies that rustdoc doesn't ICE when it encounters an IO error
+// while generating files. Ideally this would be a rustdoc-ui test, so we could
+// verify the error message as well.
+//
+// It operates by creating a temporary directory and modifying its
+// permissions so that it is not writable. We have to take special care to set
+// the permissions back to normal so that it's able to be deleted later.
+
+use run_make_support::{rustdoc, tmp_dir};
+use std::fs;
+
+fn main() {
+    let out_dir = tmp_dir().join("rustdoc-io-error");
+    let output = fs::create_dir(&out_dir).unwrap();
+    let mut permissions = fs::metadata(&out_dir).unwrap().permissions();
+    let original_permissions = permissions.clone();
+    permissions.set_readonly(true);
+    fs::set_permissions(&out_dir, permissions.clone()).unwrap();
+
+    let output = rustdoc().input("foo.rs").output(&out_dir).command_output();
+
+    // Changing back permissions.
+    fs::set_permissions(&out_dir, original_permissions).unwrap();
+
+    // Checks that rustdoc failed with the error code 1.
+    #[cfg(unix)]
+    assert_eq!(output.status.code().unwrap(), 1);
```

Comment:
Hmm, do you think we should just have a check on Windows that the error code is not 101 (i.e. ICE)? It doesn't really matter what the specific exit code is, it just has to not ICE (which by proxy not 101)?

Reply:
Also, maybe we want to print the ICE message in the case that the assertion *does* fail to aid debugging?